### PR TITLE
Use zero page "holes" to optimize sparse byte array usage

### PR DESCRIPTION
**Pull request by original-brownbear**
Add the notion of a "zero page" or "hole" to big arrays. We have some use cases where we run up byte arrays of hundreds of MB that are extremely sparse.
Each page starts out as a "hole" and only gets replaced by a real page from the pool on write similar to how FS holes work.
This change adds a small amount of overhead to the write side but is performance neutral or better on the read side (for sparse arrays we likely get a big improvement from using less CPU cache).

The only change outside of the array itself this needed was in CCR, see inline comments for that.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/common/util/AbstractBigByteArray.java [hunk 1]
```diff
@@ -10,21 +10,31 @@
...
     protected byte[][] pages;
 
     protected AbstractBigByteArray(int pageSize, BigArrays bigArrays, boolean clearOnResize, long size) {
         super(pageSize, bigArrays, clearOnResize);
         this.size = size;
         pages = new byte[numPages(size)][];
-        for (int i = 0; i < pages.length; ++i) {
-            pages[i] = newBytePage(i);
+        Arrays.fill(pages, ZERO_PAGE);
+        assert assertZeroPageClean();
```
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/common/util/AbstractBigByteArray.java:30]: Had some bugs with this, it's a little wasteful but having this in place lets me sleep better I think.

File: server/src/main/java/org/elasticsearch/common/util/AbstractBigByteArray.java [hunk 2]
```diff
@@ -68,22 +79,40 @@ protected static void fillBySelfCopy(byte[] page, int fromBytes, int toBytes, in
...
                 copyLen = Math.min(len, pageSize());
-                System.arraycopy(buf, offset << shift, pages[pageIndex], 0, copyLen << shift);
+                System.arraycopy(buf, offset << shift, getPageForWriting(pageIndex), 0, copyLen << shift);
             } while (len > copyLen);
         }
     }
+
+    protected byte[] getPageForWriting(int pageIndex) {
+        byte[] foundPage = pages[pageIndex];
+        if (foundPage == ZERO_PAGE) {
```
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/common/util/AbstractBigByteArray.java:103]: This is the new overhead we add to the `set` path relative to not having this. It seems to me, that it's a trivial overhead, especially when considering that we're fine with lots of redundant reads in `set` today. Plus, I've never seen `set` hot in profiling.

File: server/src/main/java/org/elasticsearch/common/util/BigArrays.java [hunk 3]
```diff
@@ -162,8 +164,8 @@ public BytesRef next() {
         }
 
         @Override
-        public void fillWith(StreamInput in) throws IOException {
-            in.readBytes(array, 0, Math.toIntExact(size()));
+        public void fillWith(InputStream in) throws IOException {
```
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/common/util/BigArrays.java:167]: See CCR changes, we were doing a really hacking thing there reading into a `BytesReference` via it's iterator over `BytesRef`. Just passing the `ByteArray` through the CCR code and reading via this method.

**Review by nik9000 — APPROVED** [review]
Seems worth a shot!

**Comment by original-brownbear** [comment]
Thanks Nik, 🤞 :)

**Pull request merged**